### ISSUE TITLE
skip transpose optimize

### ISSUE THREAD

**Issue by ArtyZe** [onnx/tensorflow-onnx#868]
Thanks for your job, but the transpose optimizer makes me confused because it is very very complicate. The question is so easy that I need to convert input data format from NHWC to NCHW, I can just transpose the image as pre process instead of creat so many transpose operators and then delete them one by one.  I  changed lots of codes to realize above idea, and it runs faster and becomes easy to read

**Comment by jignparm** [comment]
Are you able to convert the usual pre-trained models using this approach? The major problem is not just switching memory layout of the initial tensor-- it's converting  all the downstream TF ops with attributes, inputs, weights etc. (trained assuming NHWC), into ONNX memory layout (which assumes NCHW.

Try it out for any resnet or mobilnet models, and please update this thread if it works!

**Comment by ArtyZe** [comment]
@jignparm 
Yes, I changed lots of codes for example, Biasadd, Concat, especially these two function"conv_convert_inputs" and "_optimize_at_current_graph_level" to skip transpose.
As far it works fine in resnet, mobilenet... except for seresnet because split op cannot split the input tensor correctly. And the main reason is the transpose operator added before conv too.
In brief, If you want to let this repo concept by more and more people, please just remove the mess transpose because it's really not important for convert and not good for us to understand your great work

**Comment by jignparm** [comment]
The transpose optimizer has been present for a while -- I'm assuming the changes were incremental as more models were optimized, which may be why the code looks convoluted.

It would be good to understand which portions can be trimmed however.  Main objective would be to not lose any functionality, and still end up with maximally-optimized ONNX models.

**Comment by ArtyZe** [comment]
got it. I don’t mean to judge your work but kindly advice. But after optimize your code, it works exactly efficient and easy to understand. I think it worth to remove the mess transpose nodes before and after conv nodes, just put them in pre and post process.
I can upload my code to creat a new repo in git, you can make a comparison between them
| |
高扬
|
|
邮箱：gaoyang917528@163.com
|

签名由 网易邮箱大师 定制

On 04/02/2020 18:11, jignparm wrote:

The transpose optimizer has been present for a while -- I'm assuming the changes were incremental as more models were optimized, which may be why the code looks convoluted.

It would be good to understand which portions can be trimmed however. Main objective would be to not lose any functionality, and still end up with maximally-optimized ONNX models.

**Comment by guschmue** [comment]
If there is a better way of doing this we'd certainly consider changing. We have not had issues with transposes for some time now so we did not spend much time thinking about it.

The thought behind the current code goes like this:

The last thing we want are numerical errors so we need to be absolutely sure we don't miss one of those data_format transforms. 
In each op that has a data_format we call conv_convert_inputs() at the end to insert the needed transposes. There are 9 ops that need that.
conv_convert_inputs() is in theory easy and just inserts the transposes if needed. In practice it is a little more complicated because if the input is a const we'll transpose the const instead of having a extra op and kernels need a little different transpose. Now that I look at this again we should have called it different. 
Instead of calling conv_convert_inputs() from the op handler itself we could make a separate pass over the entire graph as rewriter but that would move things without making it much easier to understand.

Once we are done with the graph the optimizer goes over the entire graph. It tries to push transposes down in the graph until it bumps into another transpose. If that transpose is the inverse those 2 transposes cancel each other out, if not we have end up with 1 transpose.
It is complicated because for some ops we need to be aware of how to push the transpose over that op. For example for slice:  x[3] in NHWC needs to become x[1] in NCHW. 
This has been working pretty ok for us. Most issues we saw is that there is an op where we have not added the handler that knows how to deal with that specific op. 

We certainly appreciate if you show us a better or easier  way of doing this.
